libsweep/src/protocol.cc

```diff
@@ -55,91 +55,100 @@ void write_command_with_arguments(serial::device_s serial, const uint8_t cmd[2],
   serial::device_write(serial, &packet, sizeof(cmd_param_packet_s));
 }
 
-void read_response_header(serial::device_s serial, const uint8_t cmd[2], response_header_s* header) {
+response_header_s read_response_header(serial::device_s serial, const uint8_t cmd[2]) {
   SWEEP_ASSERT(serial);
   SWEEP_ASSERT(cmd);
-  SWEEP_ASSERT(header);
 
-  serial::device_read(serial, header, sizeof(response_header_s));
+  response_header_s header;
+  serial::device_read(serial, &header, sizeof(header));
 
-  uint8_t checksum = checksum_response_header(*header);
+  uint8_t checksum = checksum_response_header(header);
 
-  if (checksum != header->cmdSum)
+  if (checksum != header.cmdSum)
     throw error{"invalid response header checksum"};
 
-  bool ok = header->cmdByte1 == cmd[0] && header->cmdByte2 == cmd[1];
+  bool ok = header.cmdByte1 == cmd[0] && header.cmdByte2 == cmd[1];
 
   if (!ok)
     throw error{"invalid header response commands"};
+
+  return header;
 }
 
-void read_response_param(serial::device_s serial, const uint8_t cmd[2], response_param_s* param) {
+response_param_s read_response_param(serial::device_s serial, const uint8_t cmd[2]) {
   SWEEP_ASSERT(serial);
   SWEEP_ASSERT(cmd);
-  SWEEP_ASSERT(param);
 
-  serial::device_read(serial, param, sizeof(response_param_s));
+  response_param_s param;
+  serial::device_read(serial, &param, sizeof(param));
 
-  uint8_t checksum = checksum_response_param(*param);
+  uint8_t checksum = checksum_response_param(param);
 
-  if (checksum != param->cmdSum)
+  if (checksum != param.cmdSum)
     throw error{"invalid response param header checksum"};
 
-  bool ok = param->cmdByte1 == cmd[0] && param->cmdByte2 == cmd[1];
+  bool ok = param.cmdByte1 == cmd[0] && param.cmdByte2 == cmd[1];
 
   if (!ok)
     throw error{"invalid param response commands"};
+
+  return param;
 }
 
-void read_response_scan(serial::device_s serial, response_scan_packet_s* scan) {
+response_scan_packet_s read_response_scan(serial::device_s serial) {
   SWEEP_ASSERT(serial);
-  SWEEP_ASSERT(scan);
 
-  serial::device_read(serial, scan, sizeof(response_scan_packet_s));
+  response_scan_packet_s scan;
```

Comment:
To nitpick... we use the term "scan" frequently to mean a collection of scan_packets. So to avoid confusion I would change this variable name.  Perhaps to `packet`.